File: 285. Inorder Successor in BST.cpp
//Solution1
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* inorderSuccessor(TreeNode* root, TreeNode* p) {
        TreeNode* prev = NULL, *cur = root, *pred, *res = NULL;
        while (cur) {
            if (!cur->left) {
                if (prev == p) {
                    res = cur;
                }
                prev = cur;
                cur = cur->right;
            } else {
                pred = cur->left;
                while (pred->right != NULL && pred->right != cur) {
                    pred = pred->right;
                }
                if (pred->right == NULL) {
                    pred->right = cur;
                    cur = cur->left;
                } else if (pred->right == cur) {
                    pred->right = NULL;
                    if (pred == p) {
                        res = cur;
                    }
                    prev = cur;
                    cur = cur->right;
                }
            }
        }
        return res;
    }
};

//Solution2
TreeNode* inorderSuccessor(TreeNode* root, TreeNode* p) {
    TreeNode *successor = NULL;
    while (root) {
        if (p->val < root->val) {
            successor = root;
            root = root->left;
        } else {
            root = root->right;
        }
    }
    return successor;
}

//Predecessor
TreeNode* inorderPredecessor(TreeNode* root, TreeNode* p) {
    TreeNode *predecessor = NULL;
    while (root) {
        if (p->val > root->val) {
            predecessor = root;
            root = root->right;
        } else {
            root = root->left;
        }
    }
    return predecessor;
}

File: java/285. Inorder Successor in BST.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
// O(logn)
class Solution {
    public TreeNode inorderSuccessor(TreeNode root, TreeNode p) {
        if (root == null || p == null) {
            return null;
        }
        TreeNode suc = null;
        if (p.right != null) {
            suc = p.right;
            while (suc.left != null) {
                suc = suc.left;
            }
            return suc;
        }
        while (root != null && root != p) {
            if (root.val < p.val) {
                root = root.right;
            } else {
                suc = root;
                root = root.left;
            }
        }
        return suc;
    }
}
